components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,368 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use rusqlite::types::ToSqlOutput;
+use rusqlite::{named_params, Result as RusqliteResult, ToSql};
+use sql_support::ConnExt;
+use url::form_urlencoded;
+
+use crate::{
+    db::SuggestDao,
+    provider::SuggestionProvider,
+    rs::{DownloadedYelpSuggestion, SuggestRecordId},
+    suggestion::Suggestion,
+    Result, SuggestionQuery,
+};
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[repr(u8)]
+enum Modifier {
+    PreModifier = 0,
+    PostModifier = 1,
+}
+
+impl ToSql for Modifier {
+    fn to_sql(&self) -> RusqliteResult<ToSqlOutput<'_>> {
+        Ok(ToSqlOutput::from(*self as u8))
+    }
+}
+
+/// This module assumes like following query.
+/// "Pre-modifier? Subject Post-modifier? (Location-modifier | Location-sign Location?)?"
+/// For example, the query below is valid.
+/// "Best(Pre-modifier) Ramen(Subject) Delivery(Post-modifier) In(Location-sign) Tokyo(Location)"
+/// Also, as everything except Subject is optional, "Ramen" will be also valid query.
+/// However, "Best Best Ramen" and "Ramen Best" is out of the above appearance order rule,
+/// parsing will be failed. Also, every words except Location needs to be registered in DB.
+/// Please refer to the query test in store.rs for all of combination.
+/// Currently, the maximum query length is determined while refering to having word lengths in DB
+/// and location names.
+/// max subject: 50 + pre-modifier: 10 + post-modifier: 10 + location-sign: 7 + location: 50 = 127 = 150.
+const MAX_QUERY_LENGTH: usize = 150;
+
+/// The max number of words consisting the modifier. To improve the SQL performance by matching with
+/// "keyword=:modifier" (please see is_modifier()), define this how many words we should check.
+const MAX_MODIFIER_WORDS_NUMBER: usize = 2;
+
+impl<'a> SuggestDao<'a> {
+    /// Inserts the suggestions for Yelp attachment into the database.
+    pub fn insert_yelp_suggestions(
+        &self,
```

Comment:
```suggestion
        &mut self,
```

This method writes to the database, so let's make it `mut` to match our other write methods.